src/k8s-engine/k3sHelper.ts

```diff
@@ -461,6 +479,92 @@ export default class K3sHelper extends events.EventEmitter {
     checksum: { current: 0, max: 0 },
   };
 
+  /**
+   * Find the cached version closest to the desired version.
+   * @param desiredVersion The semver of the version of k3s the system would prefer to use, with a '+k3s###' suffix
+   * @returns A semver of the version to use, also with a '+k3s###' suffix
+   */
+  async selectClosestImage(desiredVersion: semver.SemVer): Promise<semver.SemVer> {
+    const cacheDir = path.join(paths.cache, 'k3s');
+    const k3sFilenames = (await fs.promises.readdir(cacheDir))
+      .filter(dirname => /^v\d+\.\d+\.\d+\+k3s\d+$/.test(dirname));
+
+    return this.selectClosestSemVer(desiredVersion, k3sFilenames);
+  }
+
+  /**
+   * Given a semver for the desired version, and a list of directory names representing other
+   * k3s versions (matching /v\d+\.\d+\.\d+\+k3s\d+/), return the semver for the directory name
+   * that is considered closest to the desired version:
+   *
+   * Precondition: the desired version wasn't found.
+   * Return the oldest version newer than the desired version.
+   *   If there is more than one, favor the one with the highest 'k3s' value
+   * Otherwise return the newest version older than the desired version.
+   * Otherwise throw a `NoCachedK3sVersionsError` error.
+   * @param desiredVersion: a semver for the version currently specified in the config
+   * @param k3sFilenames: a list of existing cache directories in the cache directory.
+   */
+  protected selectClosestSemVer(desiredVersion: semver.SemVer, k3sFilenames: Array<string>): semver.SemVer {
+    if (k3sFilenames.length === 0) {
+      throw new NoCachedK3sVersionsError();
+    }
+    const existingVersions = k3sFilenames.map(filename => new semver.SemVer(filename));
+
+    existingVersions.sort((v1, v2): number => {
+      const diff = semver.compare(v1, v2);
+
+      return diff !== 0 ? diff : this.compareBuildVersions(v1, v2);
+    });
+    const filteredVersions = this.keepHighestBuildVersion(existingVersions);
+
+    for (let i = 0; i < filteredVersions.length; i++) {
+      const diff: number = semver.compare(filteredVersions[i], desiredVersion);
+
+      if (diff >= 0) {
+        return filteredVersions[i];
+      }
+    }
```

Comment:
It doesn't need to take build numbers into account because multiple version-hits are filtered out right before the loop. As for the comparator, you're suggesting the wrong one. `filteredVersions.find(v => v.compare(desiredVersion) >= 0)` is good enough (because versions that differ only by build versions have been filtered out).